MM-18107: Collapsed Reply Threads, Global Threads

calebroseland opened the pull request:
See also: #7744, as a predecessor to this. (Diff here: https://github.com/mattermost/mattermost-webapp/compare/crt-precursor-supporting...feature/collapsed-reply-threads)

#### Summary

- [MM-30302](https://mattermost.atlassian.net/browse/MM-30302): Global Threads screen
~~- [MM-32019](https://mattermost.atlassian.net/browse/MM-32019): Feature flag detection~~ now in [webapp's redux](https://github.com/mattermost/mattermost-webapp/blob/feature/collapsed-reply-threads/packages/mattermost-redux/src/selectors/entities/preferences.ts#L244-L277)
- [MM-30303](https://mattermost.atlassian.net/browse/MM-30303): System Console setting
- [MM-30305](https://mattermost.atlassian.net/browse/MM-30305): Account Settings toggle
- IDs removed from `RhsThread`/`ThreadViewer`: 
    | Original | E2E Equivalent | Used in styling selectors? |
    | - | - | - | 
    | `#rhsContent` | `#rhsContainer .post-right__content` | No 
    |`#rhsPostList` | `#rhsContainer .post-right-comments-container` | No
    | `#rhsFooter` | `#rhsContainer .post-create` | No


Most of this feature is safely behind the flag `MM_FEATUREFLAGS_COLLAPSEDTHREADS=true`.
 
However, some necessary component abstractions/refactoring affect codepaths not behind the flag. Most notably:
- `<ThreadViewer/>` (the guts of `<RhsThread/>`) needs to be tested extensively (non-feat. UX flows)
- ~~`<RHSSearchNav/>` (search bar, at-mentions, saved posts, help menu)~~ (this was extracted to #7744)

The bulk of the UI is in [components/threading](https://github.com/mattermost/mattermost-webapp/tree/feature/collapsed-reply-threads/components/threading)
```
components/threading
├── channel_threads/thread_footer
├── common
│   ├── button
│   ├── follow_button
├── global_threads
│   ├── thread_item
│   ├── thread_list
│   ├── thread_menu
│   ├── thread_pane
│   ├── global_threads (main screen where above comes together)
├── global_threads_link
├── thread_viewer
```

Prior versions of some of these components have been already been reviewed, but there have been some changes since. (https://github.com/mattermost/mattermost-webapp/pull/6585)


#### Screenshots
<img width="1323" alt="CleanShot 2021-02-09 at 13 07 24@2x" src="https://user-images.githubusercontent.com/11724372/107414567-cde85380-6ad7-11eb-89ff-a65406319878.png">
<img width="616" alt="CleanShot 2021-02-09 at 13 06 55@2x" src="https://user-images.githubusercontent.com/11724372/107414574-cfb21700-6ad7-11eb-871f-9fd9bd900564.png">
<img width="616" alt="CleanShot 2021-02-09 at 13 07 06@2x" src="https://user-images.githubusercontent.com/11724372/107414583-d2147100-6ad7-11eb-9d3c-1bf84e91516a.png">

matthewbirtch reviewed (changes requested):
Awesome work on this @calebroseland a few things I found in my first round of testing. I'm unsure what pieces of the puzzle aren't implemented yet, so maybe these issues are non-issues when it's all merged together.

1. The mention badge on the 'Threads' sidebar item does not go away when I go to the global threads view

3. Mentions in threads still appear on the channel item in the sidebar. In fact, there seems to be some odd behavior with mention badges in general. Sometimes the badges don't show on the Threads sidebar item until I've clicked it. 

4. The mention badge in the global threads inbox view also does not go away once I view the thread.

5. The number of replies shown on a thread in the global threads inbox view is not updating in real-time

6. The time stamp on the global threads inbox items don't seem to match the most recent reply timestamp

7. Unfollowing a thread doesn't seem to remove it from global threads

Overall, it seems like there are a number of things that are getting delayed and not updating in real-time. 

**A few other small UI issues:**

8. Can we update the alignment on the time stamp on consecutive posts so that it's right-aligned instead of left-aligned?
![image](https://user-images.githubusercontent.com/2040554/108754852-7ce35100-7514-11eb-8967-e05eead40ddd.png)

9. I got to a state where 2 unread threads display, but the 'No unread threads' empty state also showed. Seems wrong
![image](https://user-images.githubusercontent.com/2040554/108754973-a4d2b480-7514-11eb-8c67-124e240a1109.png)

10. There appears to be an issue with timestamps displaying in compact mode in the global threads view
![image](https://user-images.githubusercontent.com/2040554/108755476-4bb75080-7515-11eb-9e92-9c0a1820bb39.png)

11. The illustrations for the empty states aren't using theme colors. I believe there is a way that @nickmisasi used with his work on cloud to ensure the svgs are using theme colors.
![image](https://user-images.githubusercontent.com/2040554/108753605-f417e580-7512-11eb-9e1b-5baf83a75afb.png)

calebroseland commented:
Note on demo server: it's not currently reflecting lates mattermost-server changes for some reason. You may need to pull locally. Latest version of server and redux have necessary infra.

calebroseland commented:
Thanks for checking things out, @matthewbirtch! (cc @reflog Matt's comments above)

I believe 2–5 probably fall under websocket events, and these should have improved a lot recently, though I think Eli's still working on a few things.

On 1, is it expected that the mentions count to go away before viewing any threads? (Cleared just by going to global threads?)

6: fixed (disappears after deselection—we should probably review those selection/unread filter UX states)

On the UI items:
9: fixed
11: fixed
8, 10: still working on

matthewbirtch commented:
> On 1, is it expected that the mentions count to go away before viewing any threads? (Cleared just by going to global threads?)

Sorry, no. I think we should clear a mention once you've actually read the thread

metanerd commented:
/update-branch

lieut-data reviewed (changes requested):
@calebroseland, I'm slightly worried about merging a feature branch of this size directly into `master`. If anything goes wrong, we'll be in the awkward position of having to revert the whole thing or play whack-a-mole around our release process.

Is it feasible to break this up into changes to existing components and the core, new functionality? (I realize the size of the ask here...)

stylianosrigas commented:
Removing cloud test server for maintenance reasons. Please feel free to add the label again if you need it.

zefhemel commented:
To confirm here also what I agreed with @ogi-m: before merging this code (in whatever shape, probably split into smaller ones — but same will apply) the QA focus is on testing this doesn't regress anything with the feature flag **disabled**. Ideally there's also testing for this with the flag enabled, but it's essential there's no regression of the system with it disabled.

matthewbirtch commented:
@calebroseland I just happened to be playing with the threads PR and there were a few things I just noticed that I thought I should call out. Not sure if these are known issues:

1. the 'No followed threads yet' block is not horizontally centered when the RHS is open.
![image](https://user-images.githubusercontent.com/2040554/111171928-c22af980-857b-11eb-9747-93115b40d336.png)

2. If I have one thread in my Global Threads view, and then unfollow it, I get an infinite 'Loading' state
![image](https://user-images.githubusercontent.com/2040554/111172609-585f1f80-857c-11eb-990c-4494fd41ea36.png)

3. If I mark an unfollowed thread as unread from the channel view, it doesn't refollow that thread. It should as per the [spec](https://mattermost.atlassian.net/wiki/spaces/Threads/pages/1000669195/Desktop+Webapp+User+Experience#Following%2FUnfollowing-Threads)

calebroseland commented on components/channel_header_mobile/channel_header_mobile.js:
```diff
@@ -90,29 +90,36 @@ export default class ChannelHeaderMobile extends React.PureComponent {
+                                    />
+                                )}
+                            </div>
+                        ) : (
```
test

calebroseland replied:
test

calebroseland commented:
> 1. the 'No followed threads yet' block is not horizontally centered when the RHS is open.
I think this should be deferred under future responsiveness improvements (collapsing list view, etc.)

> 2.  If I have one thread in my Global Threads view, and then unfollow it, I get an infinite 'Loading' state
Fixed

> 3. If I mark an unfollowed thread as unread from the channel view, it doesn't refollow that thread. It should as per the spec
I think this should be done as a smaller piece of the in-channel experience story.

lieut-data reviewed (changes requested):
Blocking until smaller PR pulling out some of these changes is merged.

angeloskyratzakos commented:
@calebroseland There were some panics with that cloud workspace from `plugin/hclog_adapter.go:54`

lieut-data commented:
@calebroseland, if you wanted to merge `crt-precursor-supporting` into this branch (maybe you have already?) we could get started on the review while QA Review completes on the other?

Willyfrog commented:
I'm still doing the review, one thing I noticed is that if the mention is in a reply it doesn't appear in the channel, is this expected?

<img width="812" alt="Screen Shot 2021-03-25 at 19 35 28" src="https://user-images.githubusercontent.com/1515906/112529928-22087800-8da6-11eb-9638-157a5d706b15.png">

lieut-data commented on components/sidebar_right/sidebar_right.jsx:
```diff
@@ -151,10 +151,11 @@ export default class SidebarRight extends React.PureComponent {
             isPinnedPosts,
             postRightVisible,
+
+            // postRightSameAsSelectedThread,
```
Unused?

lieut-data commented on components/sidebar_right/sidebar_right.jsx:
```diff
@@ -163,6 +164,10 @@ export default class SidebarRight extends React.PureComponent {
             isExpanded,
         } = this.props;
 
+        /* if (postRightSameAsSelectedThread) {
```
Unused?

lieut-data commented on components/threading/channel_threads/thread_footer/thread_footer.tsx:
```diff
@@ -90,18 +91,19 @@ function ThreadFooter({
-                follow={follow}
-                unFollow={unFollow}
+                className='separated'
+                onClick={useCallback(() => {
```
I haven't seen this pattern in hooks before -- declaring one inline. Wondering if it makes more sense to lift it to the top of the method instead for clarity, especially if the rendering logic one days unexpectedly changes to `return;` early?

lieut-data commented on components/threading/common/button/button.tsx:
```diff
@@ -11,6 +11,7 @@ type Props = {
     append?: ReactNode;
     isActive?: boolean;
     hasDot?: boolean;
+    clamp?: boolean;
```
It's not super obvious what `clamp` means in the context of a button -- is there a better name we could use? Or do we need a comment here?

calebroseland replied:
Maybe `clampText` and a comment?

lieut-data replied:
`allowTextOverflow`, at least, now tells me what the intent is here. (Wasn't sure until now ;p)

lieut-data commented on components/threading/common/options.ts:
```diff
@@ -1,12 +1,18 @@
-    'week',
-];
+export const THREADING_TIME: Partial<ComponentProps<typeof Timestamp>> = {
+
```
nit: extra line?

calebroseland replied:
slip of the keyboard

lieut-data commented on components/threading/global_threads/index.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export {default} from './global_threads';
```
Same comment as earlier: do we need this `index.ts` at all?

lieut-data commented on components/timestamp/timestamp.tsx:
```diff
@@ -278,7 +284,8 @@ class Timestamp extends PureComponent<Props, State> {
                 const {
                     display,
-                    updateIntervalInSeconds,
+                    updateIntervalInSeconds = this.props.updateIntervalInSeconds,
```
I'm not familiar with the `autoRange` pattern here, but can you elaborate on why this line is changing?

calebroseland replied:
(`autoRange` just determines which `RangeDescriptor` to use.)

This line specifically provides a default `updateIntervalInSeconds` from props if the range descriptor doesn't provide one itself.

(The fallback to both of these are statically defined intervals based on the display unit.)

lieut-data commented on components/timestamp/timestamp.tsx:
```diff
@@ -343,6 +351,14 @@ class Timestamp extends PureComponent<Props, State> {
 
+    static getDerivedStateFromProps(props: Props, state: State) {
+        if (props.value !== state.prevValue) {
+            return ({now: new Date(), prevValue: props.value});
```
Can you elaborate on the need to touch `now` here? Does this have side effect on other uses of this component?

calebroseland replied:
Most of the time `value` is older than `now` e.g. a fixed timestamp with an ongoing `now` that updates on an interval.

But when a new `value` comes down the pipe, this mechanism ensures that `now` isn't stale, preventing incorrect diffs.

I think I could and probably should move this to a separate PR.

lieut-data commented on components/user_profile/index.ts:
```diff
@@ -13,14 +13,11 @@ type OwnProps = {
-            displayName: getDisplayName(state, ownProps.userId, true),
-            user: getUser(state, ownProps.userId),
-        };
+    return {
```
Removing the callback return means `makeGetDisplayName` is created from scratch every time, breaking the intended memoization here, and we should just use the underlying, unmemoized selector instead. Or was this change intentional?

calebroseland replied:
Unintentional. I think I forgot to undo some experimentation I was doing with `useSelector`+`createSelector` and the make-selector pattern re: `makeGetDisplayName`. I'll revert

lieut-data reviewed (changes requested):
I've gotten through about 20% of the review, but figured I'd push up my comments now to help surface feedback early.

Also, I realized that by only merging the broken out PR back into this one, all the "deltas" are still present and hard to ignore using GitHub's review interface. Propose we continue with developer review here once the other PR is merged and master gets pulled into this branch, narrowing the scope of the files in question considerably.

calebroseland commented:
@Willyfrog, there's a successor PR that @reflog is working on to make changes to the mention badging, so that is okay for now. There's a spec I'll send you a link to that describes it in more detail.

Willyfrog commented on components/user_settings/display/index.ts:
```diff
@@ -54,6 +54,8 @@ function mapStateToProps(state: GlobalState) {
         channelDisplayMode: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.CHANNEL_DISPLAY_MODE, Preferences.CHANNEL_DISPLAY_MODE_DEFAULT),
         messageDisplay: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.MESSAGE_DISPLAY, Preferences.MESSAGE_DISPLAY_DEFAULT),
         collapseDisplay: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.COLLAPSE_DISPLAY, Preferences.COLLAPSE_DISPLAY_DEFAULT),
+        collapsedReplyThreadsAllowUserPreference: isCollapsedThreadsAllowed(state) && getConfig(state).CollapsedThreads as string !== 'always_on',
```
nit: 'always_on' as a constant

calebroseland replied:
CollapsedThreads is also backed by explicit string literals, but we could possibly change it to an enum. I'll go ahead and do that, but maybe I'll queue up the topic for the next guild meeting? (Constants, enums, string literal unions, etc.)

Willyfrog commented on components/widgets/header/header.tsx:
```diff
@@ -0,0 +1,41 @@
+
+type Props = {
+    heading: ReactNode;
+    level?: 0 | 1 | 2 | 3 | 4 | 5 | 6;
```
this should be a  named type, I'm guessing this could be reused in future components

calebroseland replied:
I'm a huge fan of reusability, but I think `level` is fairly specific to `Headings` below, though it's still accessible via `React.ComponentProps<typeof Header>['level']` if need be, or I could export the props type?

Willyfrog commented on stores/hooks.ts:
```diff
@@ -0,0 +1,49 @@
+    suffix: string = useSelector(currentUserAndTeamSuffix),
+): [TVal, (value: TVal) => ReturnType<typeof setGlobalItem>] {
+    const dispatch = useDispatch();
+    const storedKey = name + suffix;
```
nit
```suggestion
    const storedKey = `${name}${suffix}`;
```

Willyfrog reviewed (changes requested):
the new feature looks awesome! I'm really looking forward to have that in daily :D

I've added a few comments, but it looks great in general.

You might also want to look into views/channel_sidebar.ts (I think that's the place, but might be mistaken) to change the mouse pointer to a prohibited sign when dragging a channel on top of the globalThreadsLink component

another thing i noticed, on a long thread (not fitting the screen) if I press enter I get a different scroll than if I start writing (the add comment button doesn't show in the first case)

also no drag'n drop file, but I guess this might be for the future?

calebroseland commented on actions/websocket_actions.jsx:
```diff
@@ -1359,3 +1375,46 @@ export function handleUserActivationStatusChange() {
+function handleThreadReadChanged(msg) {
+    return (doDispatch, doGetState) => {
+        if (msg.data.thread_id) {
+            const thread = doGetState().entities.threads.threads?.[msg.data.thread_id];
```
yes, it's feature specific. I could see us using  `userThreads.threads`, but it already went into the redux code in prior PRs. The feature flag will still be off and ongoing changes will be made, so I think this should be deferred for another time/follow up PR.

calebroseland commented on actions/websocket_actions.jsx:
```diff
@@ -1359,3 +1375,46 @@ export function handleUserActivationStatusChange() {
+            const threadData = JSON.parse(msg.data.thread);
+            handleThreadArrived(doDispatch, doGetState, threadData, msg.broadcast.team_id);
+        } catch {
+            // invalid JSON
```
Probably. This was a part of initial unreads/mentions. @reflog thoughts? is this catch still needed or can we remove it?

calebroseland commented on components/admin_console/admin_definition.jsx:
```diff
@@ -5609,6 +5609,46 @@ const AdminDefinition = {
+                                display_name_default: 'Disabled',
+                            },
+
+                            /* {
+                                value: 'off',
+                                display_name: t('admin.experimental.collapsedThreads.off'),
+                                display_name_default: 'Disabled',
+                            }, */
```
`off` was originally in the spec, but ended up as `disabled`, so I think we can remove. The `always_on` is for future use.

calebroseland commented on components/threading/global_threads/global_threads.tsx:
```diff
@@ -0,0 +1,189 @@
+                        setFilter={setFilter}
+                        someUnread={Boolean(numUnread)}
+                    >
+                        {(filter === 'unread' ? unreadThreadIds : threadIds).map((id) => (
```
`filter` is backed by `ThreadFilter`, so it's typed and all. I feel it's simpler to use string literal unions for smaller/local things like this.

calebroseland commented on components/threading/global_threads/global_threads.tsx:
```diff
@@ -0,0 +1,189 @@
+                })}
+                right={<RHSSearchNav/>}
+            />
+            {isEmpty(threadIds) ? (
+                <div className='no-results__holder'>
+                    {isLoading ? (
+                        <LoadingScreen/>
+                    ) : (
```
I do agree that large render trees can be a bit difficult to read, and I think this is around the limit of what should be done in a single component—and quite possibly it should be broken down into smaller chunks.

But even with trees like this one, I like the consistency in knowing that _this is the tree_—its internal forks are all here and I know that there aren't any separate fragments that I would need to go hunting for. If it should be a separate chunk/fragment, perhaps it should be its own component.

I did want to maintain a relatively flat hierarchy (e.g. header is generic, with the page-specific content here), but I'm completely open to changes—especially for readability—@Willyfrog @lieut-data what do you guys think?

Willyfrog replied:
if we are nesting `if`s for me that's a bit too much logic into what should only be rendering.

calebroseland replied:
I can refactor this into smaller pieces, and will, but I believe we should leverage the feature flag and make changes in a follow up PR given this late-stage in the process. (Beta and QA urgency with the next cloud branch cut on Monday.)

calebroseland commented on components/sidebar_right/sidebar_right.jsx:
```diff
@@ -163,6 +164,10 @@ export default class SidebarRight extends React.PureComponent {
             isExpanded,
         } = this.props;
 
+        /* if (postRightSameAsSelectedThread) {
+            return null;
+        } */
```
Past workarounds, will remove

calebroseland commented on components/markdown/index.ts:
```diff
@@ -44,6 +47,11 @@ function makeMapStateToProps() {
     return function mapStateToProps(state: GlobalState, ownProps: Props) {
         const config = getConfig(state);
 
+        let channelId;
+        if (ownProps.postId) {
+            channelId = getPost(state, ownProps.postId)?.channel_id;
+        }
```
That was my first thought, but I don't think we'd want `getPost` firing for falsy postId values.

calebroseland commented on components/threading/common/follow_button/follow_button.tsx:
```diff
@@ -1,40 +1,37 @@
-    isFollowing: boolean;
-    follow: MouseEventHandler<HTMLButtonElement>;
-    unFollow: MouseEventHandler<HTMLButtonElement>;
+    isFollowing: boolean | null | undefined;
```
You're right that we don't technically need `undefined`, I'll go ahead and remove that. But when null (while thread data is loading), the follow button is disabled to prevent errant changes/toggles.

calebroseland commented on components/threading/global_threads/thread_menu/thread_menu.tsx:
```diff
@@ -1,145 +1,138 @@
+                        defaultMessage: 'Mark as unread',
+                    })}
+                    onClick={useCallback(() => {
+                        dispatch(updateThreadRead(currentUserId, currentTeamId, threadId, hasUnreads ? Date.now() : unreadTimestamp));
```
`updateThreadRead` is a single action to handle both marking "as read" and "as unread." `now` corresponds to "as read" and `unreadTimestamp` (og post time) for "as unread."

calebroseland commented on components/channel_header_mobile/channel_header_mobile.js:
```diff
@@ -90,29 +90,36 @@ export default class ChannelHeaderMobile extends React.PureComponent {
+                                isReadOnly={isReadOnly}
+                                isRHSOpen={isRHSOpen}
+                                currentRelativeTeamUrl={currentRelativeTeamUrl}
+                            />
+                        )}
+                        <ShowSearchButton/>
                         {channel && (
-                            <React.Fragment>
-                                <div className='navbar-brand'>
-                                    <MobileChannelHeaderDropdown/>
-                                    {isMuted && (
-                                        <UnmuteChannelButton
-                                            user={user}
-                                            channel={channel}
-                                        />
-                                    )}
-                                </div>
-                                <ChannelInfoButton
-                                    channel={channel}
-                                    isReadOnly={isReadOnly}
-                                    isRHSOpen={isRHSOpen}
-                                    currentRelativeTeamUrl={currentRelativeTeamUrl}
-                                />
-                                <ShowSearchButton/>
-                                <MobileChannelHeaderPlug
-                                    channel={channel}
-                                    isDropdown={false}
-                                />
-                            </React.Fragment>
+                            <MobileChannelHeaderPlug
+                                channel={channel}
+                                isDropdown={false}
+                            />
```
I'm using inline conditionals here to avoid two very similar trees with would-be duplicated content. It would be more straight forward if the items weren't interwoven, and we should probably improve this, but I was trying to reduce changes to core layout when possible.

calebroseland commented on components/favicon_title_handler/favicon_title_handler.tsx:
```diff
@@ -83,21 +84,32 @@ class FaviconTitleHandler extends React.PureComponent<Props> {
+            }, {
+                prefix: mentionTitle + unreadTitle,
+                displayName: currentTeam.display_name,
+                siteName: currentSiteName,
```
Up on :91 okay?

calebroseland replied:
@Willyfrog after a second look at this, I think we should leave it as-is. https://github.com/mattermost/mattermost-webapp/blob/master/components/favicon_title_handler/favicon_title_handler.test.jsx#L53-L56 tests specifically for a null sitename. If we do need to change it, I think this should be in a separate PR as it would change non-threading titles.

calebroseland commented on components/favicon_title_handler/index.ts:
```diff
@@ -24,6 +28,7 @@ function mapStateToProps(state: GlobalState) {
         currentTeammate,
         siteName: config.SiteName,
         unreads: getUnreads(state),
+        isGlobalThreadsView: match?.path === '/:team' && location?.pathname.includes(`${match?.url}/threads`),
```
We're outside of that route path, so match.path will only ever get to /:team here, but yes, there's a better way to check, I'll switch it

calebroseland commented on components/admin_console/admin_definition.jsx:
```diff
@@ -5609,6 +5609,46 @@ const AdminDefinition = {
+                                display_name_default: 'Disabled',
+                            },
+
+                            /* {
```
Was, but we can remove `off`, `always_on` is for future use.

calebroseland commented on actions/websocket_actions.jsx:
```diff
@@ -1359,3 +1375,46 @@ export function handleUserActivationStatusChange() {
+
+function handleThreadFollowChanged(msg) {
+    return (doDispatch) => {
+        handleFollowChanged(doDispatch, msg.data.thread_id, msg.broadcast.team_id, msg.data.state);
```
The `handle*` methods are more or less just action creators with an occasional thunk signature—there's no reason we can't remove the thunk and dispatch-passing logic. `actions/threads.ts` developed in part based on proximity and similar/copied actions.

calebroseland commented on actions/websocket_actions.jsx:
```diff
@@ -1359,3 +1375,46 @@ export function handleUserActivationStatusChange() {
+                );
+            }
+        } else {
+            handleAllMarkedRead(doDispatch, msg.broadcast.team_id);
```
@reflog would that be easily doable?

calebroseland commented on actions/websocket_actions.jsx:
```diff
@@ -1359,3 +1375,46 @@ export function handleUserActivationStatusChange() {
+function handleThreadReadChanged(msg) {
+    return (doDispatch, doGetState) => {
+        if (msg.data.thread_id) {
+            const thread = doGetState().entities.threads.threads?.[msg.data.thread_id];
```
I agree that should be possible, but I'll double check to make sure there isn't a specific reason it had to be outside of/before the action.

calebroseland commented on actions/team_actions.jsx:
```diff
@@ -105,3 +110,12 @@ export function updateTeamsOrderForUser(teamIds) {
+    return async (dispatch, getState) => {
+        const state = getState();
+        if (isCollapsedThreadsEnabled(state)) {
+            dispatch(getThreadMentionCountsByChannel(getCurrentTeamId(state)));
```
(cc @reflog in case I'm off on something or missed anything)

- `getThreadMentionCountsByChannel` is done on a per-team basis
- Looks like this is fired on WS `reconnect` and `focusPost`
- it could get it's own state and short circuit internally, but currently all feature-on-off gate logic is outside of `actions/threads.ts`
  
In the end, looks like this team action is being removed in the next PR as further changes to mentions/unreads are underway.
https://github.com/mattermost/mattermost-webapp/pull/7733/files#diff-42c7d37cdea1c44e8773b4f6a75c9c03291f9bffebb8d268186c0a41ff929a3f

calebroseland commented on components/legacy_sidebar/sidebar_channel/index.js:
```diff
@@ -62,7 +63,11 @@ function makeMapStateToProps() {
         let showUnreadForMsgs = true;
         if (member) {
-            unreadMentions = member.mention_count;
+            const threadCounts = getThreadCountsInCurrentTeam(
+                state,
+            );
```
Changing in next PR https://github.com/mattermost/mattermost-webapp/pull/7733/files#diff-b1ad3be55e1e68238de30c163be916663fb761d4493ed423aab7cc59c32b9aee

calebroseland commented on components/legacy_sidebar/sidebar_channel/index.js:
```diff
@@ -62,7 +63,11 @@ function makeMapStateToProps() {
+            const threadCounts = getThreadCountsInCurrentTeam(
+                state,
+            );
+            const threadMentionCountInChannel = (threadCounts && threadCounts.unread_mentions_per_channel && threadCounts.unread_mentions_per_channel[channel.id]) || 0;
```
@reflog ?

calebroseland commented on components/legacy_sidebar/sidebar_channel/index.js:
```diff
@@ -62,7 +63,11 @@ function makeMapStateToProps() {
+                state,
+            );
+            const threadMentionCountInChannel = (threadCounts && threadCounts.unread_mentions_per_channel && threadCounts.unread_mentions_per_channel[channel.id]) || 0;
+            unreadMentions = member.mention_count - threadMentionCountInChannel;
```
@reflog ?

matthewbirtch reviewed (approved):
Approving this now to not be a blocker and will do more robust UX testing when merged and on community-daily.

lieut-data commented on components/threading/global_threads_link/threads_icon.tsx:
```diff
@@ -0,0 +1,23 @@
+    );
+};
+
+export default memo(ThreadsIcon);
```
Do we require `memo` here? It seems like this component is simple enough that the cost of memoization may be greater than just letting React handle things.

calebroseland replied:
Memo all the things
![image](https://user-images.githubusercontent.com/11724372/112894453-7f9c0d80-90a1-11eb-8eec-9684c55af07f.png)

/s

I agree—not worth it! (will remove it)

lieut-data commented on components/widgets/users/avatar/avatar.tsx:
```diff
@@ -23,11 +27,16 @@ const Avatar = ({
             <div
                 {...attrs}
+                style={{
+                    background: tinycolor.mix(centerChannelBg, centerChannelColor, 8).toRgbString(),
```
nit: it seems unusual to compute a colour based on our constants instead of using a constant defined by the theme directly. Are we sure that these colours will always "mix" correctly?

calebroseland replied:
This was a tricky one. It's a transparent bg, but there's an overlapping object so for this one element it needed to be opaque... testing in the 4 default themes we have yielded the expected colors, but there may be edge cases.

I'll mark it down for a more thorough UI check during UX review with custom themes.

lieut-data commented on reducers/views/threads.ts:
```diff
@@ -0,0 +1,24 @@
+import {Threads} from 'utils/constants';
+
+export const selectedThreadIdInTeam = (state: ViewsState['threads']['selectedThreadIdInTeam'] | null = null, action: GenericAction) => {
+    if (action.type === Threads.CHANGED_SELECTED_THREAD) {
```
nit: can we use the `switch` block despite needing to only handle one action type?

lieut-data commented on reducers/views/threads.ts:
```diff
@@ -0,0 +1,24 @@
+    if (action.type === Threads.CHANGED_SELECTED_THREAD) {
+        return {
+            ...state,
+            [action.data.team_id]: action.data.id,
```
nit: should we be using `team_id` and `post_id` or some such to clearly explain what the `id` is?

calebroseland replied:
Even though I think it's ok when the subject is small and clear, I do think in many cases it's good to avoid generic names.

lieut-data commented on selectors/views/threads.ts:
```diff
@@ -0,0 +1,45 @@
+    return state.views.threads.selectedThreadIdInTeam;
+}
+
+export const getSelectedThreadIdInCurrentTeam: (state: GlobalState) => ViewsState['threads']['selectedThreadIdInTeam'][$ID<Team>] = createSelector(
```
Do we need to memoize this with `createSelector`? (As in, is there evidence that this selector is both called often and slow to compute?) If not, can we just make this a simple function?

calebroseland replied:
(Follow up PR?)

lieut-data commented on selectors/views/threads.ts:
```diff
@@ -0,0 +1,45 @@
+import {GlobalState} from 'types/store';
+import {ViewsState} from 'types/store/views';
+
+export function getSelectedThreadIdInTeam(state: GlobalState) {
```
Using the force, this one doesn't /feel/ like it should be exported, but that clients of the "API" here should be more specific.

calebroseland replied:
Generally speaking, I see almost all these kinds of "simple selectors" being exported around here, so it was more/less aiming to follow the pattern. Though I do agree, since it isn't used externally, there's no need to export.

lieut-data commented on selectors/views/threads.ts:
```diff
@@ -0,0 +1,45 @@
+        threads,
+    ) => {
+        const threadId = selectedThreadIdInTeam?.[currentTeamId];
+        if (!threadId) {
```
Redundant check given below?

calebroseland replied:
Looks like a bad merge, will clean it up

lieut-data commented on selectors/views/threads.ts:
```diff
@@ -0,0 +1,45 @@
+    },
+);
+
+export const getSelectedThreadInCurrentTeam: (state: GlobalState) => UserThread | null = createSelector(
```
Same question as above re: `createSelector`.

calebroseland replied:
Generally, I haven't done extensive performance testing on the lower end here with these smaller selectors, so I'm not sure if they need memoization, but I generally default to them akin to preferring `PureComponent` over `Component`.

At this late stage, I'm tempted to say that we should leave them as-is for now, and address any no-need-to-memoize issues in a follow up PR shortly afterwards that would address any performance learnings (on both the too slow/not enough and no-need ends).

lieut-data commented on stores/hooks.ts:
```diff
@@ -0,0 +1,49 @@
+import {makeGetGlobalItem} from 'selectors/storage';
+import {setGlobalItem} from 'actions/storage';
+
+export const currentUserAndTeamSuffix = createSelector([
```
Does this need to be memoized?

calebroseland replied:
(Follow up PR?)

lieut-data commented on stores/hooks.ts:
```diff
@@ -0,0 +1,49 @@
+    return `:${userId}:${teamId}`;
+});
+
+export const currentUserSuffix = createSelector([
```
Does this need to be memoized?

calebroseland replied:
(Follow up PR?)

lieut-data commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -11,34 +20,88 @@ import SimpleTooltip, {useSynchronizedImmediate} from 'components/widgets/simple
+        return null;
+    }
+
+    const getDisplayName = makeGetDisplayName();
```
This creates a memoized selector on each invocation of this method.

calebroseland replied:
I reworked this into a new selector that will only memoize the display preference, and return a memoized selector function.

lieut-data commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -11,34 +20,88 @@ import SimpleTooltip, {useSynchronizedImmediate} from 'components/widgets/simple
+    users: Props['users'],
+    participants: Props['participants'],
+): null | {users: AvatarProps[]; missingProfiles?: string[]} {
+    if (users) {
```
This /seems/ to be a form of memoization? I'm having trouble following this method.

calebroseland replied:
It's an API variance short circuit if the more-direct `AvatarProps[]` were provided instead of `Array<{id: UserId}> which requires selection`

calebroseland replied:
I opted to remove the API variance and simplify it to just user ids.

lieut-data commented on components/widgets/users/avatars/avatars.tsx:
```diff
@@ -11,34 +20,88 @@ import SimpleTooltip, {useSynchronizedImmediate} from 'components/widgets/simple
+
+    const getDisplayName = makeGetDisplayName();
+
+    return participants.reduce<NonNullable<ReturnType<typeof selectUsers>>>((result, {id}) => {
```
This returns a new object every time, which means this selector forces a React reconciliation every time, which may be a performance issue (i.e. "slow typing issue" recently worked around). But maybe that's what we're trying to avoid above with the `users` check?

Should we move all this into the render method instead, and only select the minimum necessary data from the store and do everything consistently at render time?

calebroseland replied:
The first implementation of this actually did just map in the render method, but there it isn't able to take username display preference into account. 

I knew there was a perf issue with returning an obj/arr here, but doing the selection/mapping at this level _seemed_ best at the time.

Last night I was brainstorming on another redo of this—I think I'll:
- hoist the missing profiles logic and use `getMissingProfilesByIds` higher up (API layer)
- lower/defer `userId`→`Avatar` props selection into `Avatar` and add new selector for `overflowUsers`→namesOverflow tooltip

lieut-data commented on components/widgets/header/header.tsx:
```diff
@@ -0,0 +1,41 @@
+    );
+};
+
+export default memo(Header);
```
Do we need to memoize this?

calebroseland replied:
Yes, I believe it should, given the number of props and amount of content that may be in them.

It's also my understanding that a `memo`-wrapped functional component is generally equivalent to a `PureComponent` (less the state diffing), and we do usually recommend using that over `Component`.

calebroseland replied:
After another look—you're right, we don't need this. The render props make it pointless.

lieut-data commented on components/threading/thread_viewer/thread_viewer.tsx:
```diff
@@ -0,0 +1,564 @@
+    ],
+};
+export function renderView(props: Record<string, any>) {
+    return (
```
nit: for simple methods like these, I like the reduced-line-count pattern that eliminates the `return`:
```js
export function renderView = (props: Record<string, any>) => ({
    <div ...
});
```

calebroseland replied:
(was moved/existing; didn't want to touch unnecessary)

I like that pattern, too. But I'm up on the air on this one. At least while developing, I may add and remove `const`s/`let`s as the function evolves, so switching between these two becomes somewhat burdensome IMHO..  


```tsx
const thing = () => (
    // jsx
);
```
and
```tsx
const thing = () => {
    return (
        // jsx
    );
};
```

lieut-data commented on components/threading/thread_viewer/thread_viewer.tsx:
```diff
@@ -0,0 +1,564 @@
+        const openTime = (new Date()).getTime();
+
+        this.state = {
+            windowWidth: Utils.windowWidth(),
```
Haven't read the rest of the code, but this "smells" wrong -- how can we avoid needing to compute things based on the height in this way?

calebroseland replied:
(was moved/existing; didn't want to touch unnecessary)

lieut-data commented on components/threading/thread_viewer/thread_viewer.tsx:
```diff
@@ -0,0 +1,564 @@
+            userScrolledToBottom: false,
+        };
+
+        this.rhspostlistRef = React.createRef();
```
nit: `rhsPostListRef`?

calebroseland replied:
(was moved/existing; didn't want to touch unnecessary)

lieut-data commented on components/threading/thread_viewer/thread_viewer.tsx:
```diff
@@ -0,0 +1,564 @@
+
+    private getLastPost() {
+        const childPosts = this.rhspostlistRef.current?.children;
+        return childPosts && childPosts[childPosts.length - 1];
```
Any chance this can be empty in weird cases (like posts being deleted)?

calebroseland replied:
Even if there's 0 children, it appears to be `null`-safe upstream

```tsx
        const lastPost = this.getLastPost()?.getBoundingClientRect();
        //...
        if (lastPost) {
            const userScrolledToBottom = scrollTop >= maxScroll - lastPost.height;

            this.setState({userScrolledToBottom});
        }
```

lieut-data commented on components/threading/thread_viewer/thread_viewer.tsx:
```diff
@@ -0,0 +1,564 @@
+        }
+    }
+
+    public shouldComponentUpdate(nextProps: Props, nextState: State) {
```
Are we certain `shouldComponentUpdate` is required here? Usually this is a performance improvement, but only needed if we've measured one. And if we have to compute the equality of lots of things anyway, it may be a wash. 

On the flip side, if we're using this to work around a problem of updating when we shouldn't, let's identify and tackle the root problem there.

calebroseland replied:
This was already here, and I'm not 100% sure of the history, but the only change I added was was for if the `userThread` was different.

I agree though that it may not be needed or the root problem should be addressed in a more appropriate way, though I do think any significant refactoring should be done separately.

`ThreadViewer` is really just the guts of `RHSThread` extracted to be reusable.

lieut-data commented on components/threading/thread_viewer/thread_viewer.tsx:
```diff
@@ -0,0 +1,564 @@
+    }
+
+    public scrollToBottom = (): void => {
+        if ($('.post-right__scroll')[0]) {
```
Can we avoid introducing a new JQuery dependency? We're aiming to drop JQuery in Q2 2021.

calebroseland replied:
(was moved/existing; didn't want to touch unnecessarily)

lieut-data commented on components/threading/global_threads_link/global_threads_link.tsx:
```diff
@@ -0,0 +1,82 @@
+        dispatch(getThreads(currentUserId, currentTeamId, {perPage: 5}));
+    }, [currentTeamId]);
+
+    if (!isFeatureEnabled || (unreadsOnly && !threadsMatch && !someUnreadThreads)) {
```
Looks like the feature gate doesn't stop the `dispatch` above? Should it?

calebroseland replied:
Good catch! Yes, it should

lieut-data commented on components/threading/global_threads_link/global_threads_link.tsx:
```diff
@@ -0,0 +1,82 @@
+    );
+};
+
+export default memo(GlobalThreadsLink);
```
Do we need to memoize this?

calebroseland replied:
There's enough logic in it that I believe it is worth it, but I don't have any benchmarks to prove the need.

lieut-data replied:
Propose we drop memoization in call cases until we can identify a specific problem and show that it fixes it -- I'm worried that in some cases, the memoization costs us more due to the component having to re-render too often anyway.

lieut-data commented on components/threading/global_threads/thread_pane/thread_pane.tsx:
```diff
@@ -0,0 +1,123 @@
+                    <>
+                        <Button
+                            className='Button___icon Button___large back'
+                            onClick={useCallback(() => select(), [])}
```
Can we avoid inlining hook definitions?

lieut-data commented on components/threading/global_threads/thread_list/thread_list.tsx:
```diff
@@ -0,0 +1,101 @@
+                        <Button
+                            className={'Button___large Margined'}
+                            isActive={currentFilter === ''}
+                            onClick={useCallback(() => setFilter(''), [])}
```
Let's avoid inlining hooks.

lieut-data commented on components/threading/global_threads/thread_item/thread_item.tsx:
```diff
@@ -1,69 +1,128 @@
+        }
+    }, [channel, thread?.post.channel_id]);
+
+    useEffect(() => {
```
This is duplicated from line 58?

lieut-data reviewed (changes requested):
Thanks, @calebroseland! You've tackled a huge project really well -- a few comments below from my final review.

(Also, I probably commented on a few things that amount to old code just being relocated and that I may not have picked up on during the incremental review. Just a heads up!)

calebroseland commented on components/rhs_comment/rhs_comment.jsx:
```diff
@@ -198,6 +199,7 @@ class RhsComment extends React.PureComponent {
                 eventTime={post.create_at}
                 postId={post.id}
                 location={Locations.RHS_COMMENT}
+                timestampProps={{...this.props.timestampProps, style: this.props.isConsecutivePost && !this.props.compactDisplay ? 'narrow' : undefined}}
```
Not in this case. The underlying `intl` apis make a distinction between `undefined` and defined(/falsy) values

calebroseland commented on components/create_comment/create_comment.jsx:
```diff
@@ -1233,7 +1233,6 @@ class CreateComment extends React.PureComponent {
             <form onSubmit={this.handleSubmit}>
                 <div
                     role='form'
-                    id='rhsFooter'
```
Yes, all E2E selectors relying on this were updated, but on second pass there was one remaining. Should be good now. 

Though I think I'll need to do a deeper follow up PR that addresses another set of Ids `#fileUploadButton`, `#fileUploadInput`, `#addCommentButton`, `#reply_textbox` etc.

calebroseland commented on components/no_results_indicator/index.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export {default} from './no_results_indicator';
```
We had discussed a while back about an indexed-folder pattern to contain like-files together (scss, tests, storybook stories, etc.).

calebroseland commented:
> another thing i noticed, on a long thread (not fitting the screen) if I press enter I get a different scroll than if I start writing (the add comment button doesn't show in the first case)

@Willyfrog could you elaborate a bit more on this? Is this in Global Threads? RHS? What was focused when you stated typing and/or pressed enter?

Willyfrog commented:
> > another thing i noticed, on a long thread (not fitting the screen) if I press enter I get a different scroll than if I start writing (the add comment button doesn't show in the first case)
> 
> @Willyfrog could you elaborate a bit more on this? Is this in Global Threads? RHS? What was focused when you stated typing and/or pressed enter?

sorry, I'll try to do a more detailed repro steps, anyways it is a minor thing.
1. open global threads, select a long thread (larger than the screen)
1. click on the text box
1. scroll up the thread
1. hit enter

you'll see that it is scrolled but the add comment button is not visible

repeat steps but type a character instead of enter. now the button is visible.

calebroseland commented:
> > > another thing i noticed, on a long thread (not fitting the screen) if I press enter I get a different scroll than if I start writing (the add comment button doesn't show in the first case)
> > 
> > 
> > @Willyfrog could you elaborate a bit more on this? Is this in Global Threads? RHS? What was focused when you stated typing and/or pressed enter?
> 
> sorry, I'll try to do a more detailed repro steps, anyways it is a minor thing.
> 
> 1. open global threads, select a long thread (larger than the screen)
> 2. click on the text box
> 3. scroll up the thread
> 4. hit enter
> 
> you'll see that it is scrolled but the add comment button is not visible
> 
> repeat steps but type a character instead of enter. now the button is visible.

Hmm that's strange. This is what I'm seeing: Input change always scrolls to bottom. Enter (for empty input) only scrolls to bottom if the _whole_ input is out of view. (Chrome/Safari/Desktop in Global Threads and RHS—here and on community-daily.)
![CleanShot 2021-04-06 at 16 10 50](https://user-images.githubusercontent.com/11724372/113779099-ce295780-96f2-11eb-9ad7-760a838e777b.gif)

Willyfrog commented on components/favicon_title_handler/favicon_title_handler.tsx:
```diff
@@ -83,21 +84,32 @@ class FaviconTitleHandler extends React.PureComponent<Props> {
+        } else if (currentTeam && inGlobalThreads) {
+            document.title = formatMessage({
+                id: 'globalThreads.title',
+                defaultMessage: '{prefix} Threads - {displayName} {siteName}',
```
```suggestion
                defaultMessage: '{prefix}Threads - {displayName} {siteName}',
```
also, would it be possible to just `formatMessage` the word threads and use the same formula for both? it will prevent subtle errors like this to happen.

calebroseland replied:
I did consider that (to match above), but I felt it would be better to move towards what `sidebar.team_select` is like below.

Willyfrog commented on components/threading/channel_threads/thread_footer/thread_footer.tsx:
```diff
@@ -17,15 +17,14 @@ import SimpleTooltip from 'components/widgets/simple_tooltip';
 type Props = {
-    participants: ComponentProps<typeof Avatars>['users'];
+    participants: Array<{id: string}>; // Post['participants']
     totalParticipants?: number;
```
this now seems redundant, right? we can probably just get it from the length of `participants`

calebroseland replied:
For the purposes of the Avatars tooltip, more than (some threshold) may be too many to display—like with reactions tooltips, I think the first XX are named with the rest as numbered overflow ("and 23 others..."). In the GA/polish phase `participants` will be optimized around this, but for now just shows all.

Willyfrog commented on components/threading/global_threads_link/global_threads_link.tsx:
```diff
@@ -0,0 +1,85 @@
+        }
+    }, [currentTeamId]);
+
+    if (!isFeatureEnabled || (unreadsOnly && !inGlobalThreads && !someUnreadThreads)) {
+        // hide link if feature disabled or filtering unreads and there are no unread threads
+        return null;
+    }
```
nit: shouldn't this be before the `useEffect` call?

calebroseland replied:
No, hooks need to be called in the same order every time (and before early returns)

Willyfrog reviewed (changes requested):
I used `/test threaded_post` and it seems it takes a while to load, but then I can scroll with no delay. I'm not really sure how that would work with really long threads in customer setups.

Also, while waiting for the thread to load I didn't have any sort of loading animation, but I guess this could be in a future PR.

Also, by creating a thread with that command, it didn't collapse into a single message

some more weird stuff. I created a second account and joined the `guille` team. this new user didn't see any replies, but just the initial message (screenshot #1). After that I clicked the reply button and closed the RHS, and now it showed that there were 7 replies (screenshot #2). Finally I replied and my global threads lighted as if there were messages unread, but the original user didn't get anything (screenshot #3)
<img width="1680" alt="Screen Shot 2021-04-06 at 22 21 55" src="https://user-images.githubusercontent.com/1515906/113774064-16626d00-9727-11eb-8569-7e46e59ed69c.png">
<img width="1679" alt="Screen Shot 2021-04-06 at 22 22 16" src="https://user-images.githubusercontent.com/1515906/113774084-1bbfb780-9727-11eb-8d87-7050f7ded002.png">
<img width="1680" alt="Screen Shot 2021-04-06 at 22 22 51" src="https://user-images.githubusercontent.com/1515906/113774096-1ebaa800-9727-11eb-8daa-832f6aa5b55b.png">

calebroseland commented on components/threading/global_threads/thread_pane/thread_pane.tsx:
```diff
@@ -0,0 +1,123 @@
+                    <>
+                        <FollowButton
+                            isFollowing={isFollowing}
+                            disabled={isFollowing == null}
```
The main reason it's in so many places is that isFollowing can come from either the `post` or `userThread`, and it is not guaranteed to be present in `userThread`. The disabled part is an earlier piece, but may not be needed. It was used as a momentary block for while the thread was loading (don't want to allow conflicting toggle requests being sent). I'll have to double check, but we may be able to remove in a follow up PR.

calebroseland commented on components/channel_layout/center_channel/center_channel.tsx:
```diff
@@ -76,9 +87,15 @@ export default class CenterChannel extends React.PureComponent<Props, State> {
+                                component={LazyGlobalThreads}
+                            />
+                        ) : null}
                         <Redirect to={lastChannelPath}/>
```
I dug a bit into those errors and it's related to a failed DB migration, it probably caused a number of issues. Re-creating the test server should hopefully resolve those (doing now)

calebroseland replied:
Also, there have been significant changes to mentions/unread with the feature on. These server changes have already been merged, and so this PR will throw and catch some API errors around this. 

https://github.com/mattermost/mattermost-webapp/pull/7733 is the webapp-side to those changes and will be merged (after this one) before we turn anything on. 

Re: this PR, it's understood that there are issues with mentions/unreads with the feature on—they will be fixed/developed incrementally ASAP. What's crucial is that there are no feature-off regressions.

calebroseland commented on components/favicon_title_handler/favicon_title_handler.tsx:
```diff
@@ -83,21 +84,32 @@ class FaviconTitleHandler extends React.PureComponent<Props> {
 
         const currentSiteName = siteName || '';
 
+        const mentionTitle = unreads.mentionCount > 0 ? '(' + unreads.mentionCount + ') ' : '';
```
I was tempted to make aside improvements like this, but this was how it was—it may have shifted, but I didn't want to change anything unnecessarily.

calebroseland replied:
> I used `/test threaded_post` and it seems it takes a while to load, but then I can scroll with no delay. I'm not really sure how that would work with really long threads in customer setups.
> Also, while waiting for the thread to load I didn't have any sort of loading animation, but I guess this could be in a future PR.

@Willyfrog Unfortunately this is about what we have now—opening long threads in RHS will do a blocking render until all posts have loaded+rendered.
<img width="179" alt="CleanShot 2021-04-06 at 22 03 49@2x" src="https://user-images.githubusercontent.com/11724372/113804349-fc735b00-9723-11eb-82b6-4f5e279438cc.png">
But I agree, we should make some improvements (ThreadViewer infinite scrolling with non-blocking rendering, loading indicators, etc.)
 
 
> Also, by creating a thread with that command, it didn't collapse into a single message

Full support for the in-channel experience will come in subsequent PRs. Some server/mm-redux work that Eli did got us part way there with Init channel-loading only including root posts, but a follow up PR is needed to explicitly disregard replies (right now they come in via WS events), add ThreadFooter, and a number of other items.

> some more weird stuff. I created a second account and joined the `guille` team. this new user didn't see any replies, but just the initial message (screenshot #1). After that I clicked the reply button and closed the RHS, and now it showed that there were 7 replies (screenshot #2). Finally I replied and my global threads lighted as if there were messages unread, but the original user didn't get anything (screenshot #3)

I'm hoping all of this is related to a DB migration issue. I re-created this test server, so replies in Global Threads should be working as-expected, though as previously mentioned: sidebar mentions/unread indicators are still in-progress until #7733 is merged.

Willyfrog commented:
> > > > another thing i noticed, on a long thread (not fitting the screen) if I press enter I get a different scroll than if I start writing (the add comment button doesn't show in the first case)
> > > 
> > > 
> > > @Willyfrog could you elaborate a bit more on this? Is this in Global Threads? RHS? What was focused when you stated typing and/or pressed enter?
> > 
> > 
> > sorry, I'll try to do a more detailed repro steps, anyways it is a minor thing.
> > 
> > 1. open global threads, select a long thread (larger than the screen)
> > 2. click on the text box
> > 3. scroll up the thread
> > 4. hit enter
> > 
> > you'll see that it is scrolled but the add comment button is not visible
> > repeat steps but type a character instead of enter. now the button is visible.
> 
> Hmm that's strange. This is what I'm seeing: Input change always scrolls to bottom. Enter (for empty input) only scrolls to bottom if the _whole_ input is out of view. (Chrome/Safari/Desktop in Global Threads and RHS—here and on community-daily.)
> ![CleanShot 2021-04-06 at 16 10 50](https://user-images.githubusercontent.com/11724372/113779099-ce295780-96f2-11eb-9ad7-760a838e777b.gif)

i was testing that on firefox, maybe it is a firefox-specific thing

lieut-data reviewed (approved):
Caleb, I'm content with my first review pass and Guillermo's eyes in wrapping this up, so I'm lifting my changes requested. Looking forward to seeing this live soon!

calebroseland commented:
@Willyfrog I think I've covered all the main points, with a few feature-on (flag-protected code) points aimed for follow up improvements. Were there any other feature-off concerns or points before we move on to QA?

Willyfrog reviewed (approved):
> @Willyfrog I think I've covered all the main points, with a few feature-on (flag-protected code) points aimed for follow up improvements. Were there any other feature-off concerns or points before we move on to QA?

Yeah, we can keep those for future PRs. lets see if we can get this merged soon 🚀

zefhemel commented:
Alright, looks like this is ready to merge. Let's do this as soon as the new cloud branch is cut (hopefully later today), to give it us plenty of time to ensure this doesn't regress anything before the next release.

ogi-m reviewed (approved):
Thanks Caleb! Approving from QA side 👍🏻 🎉
